File: 1-starter_project/common/mytravels.common/Services/OpenStreetMapsService.cs
using Flurl;
using Flurl.Http;
using Microsoft.Extensions.Configuration;
using Newtonsoft.Json.Linq;
using Polly;
using Polly.Retry;
using System.Globalization;
using mytravels.contract.Interfaces;

namespace mytravels.common.Services
{
    public class OpenStreetMapsService : IMapsService
    {
        private const string DefaultBaseUrl = "https://nominatim.openstreetmap.org";
        private const string DefaultUserAgent = "mytravels/1.0";

        private readonly IConfiguration _configuration;
        private readonly int _maxRetryAttempts = 2;
        private readonly AsyncRetryPolicy _policy;

        public OpenStreetMapsService(IConfiguration configuration)
        {
            _configuration = configuration ?? throw new ArgumentNullException(nameof(configuration));
            _policy = Policy
               .Handle<FlurlHttpTimeoutException>()
               .Or<FlurlHttpException>()
               .WaitAndRetryAsync(_maxRetryAttempts, i => TimeSpan.FromSeconds(Math.Pow(2, i)));
        }

        public async Task<string> GetAddressAsync(double latitude, double longitude, CancellationToken cancellationToken)
        {
            string baseUrl = _configuration.GetValue<string>("OpenStreetMapsUrl") ?? DefaultBaseUrl;
            string userAgent = _configuration.GetValue<string>("OpenStreetMapsUserAgent") ?? DefaultUserAgent;
            string lat = latitude.ToString(CultureInfo.InvariantCulture);
            string lon = longitude.ToString(CultureInfo.InvariantCulture);

            return await _policy.ExecuteAsync(async () =>
            {
                string response = await baseUrl.AppendPathSegment("reverse")
                                               .SetQueryParam("format", "jsonv2")
                                               .SetQueryParam("lat", lat)
                                               .SetQueryParam("lon", lon)
                                               .WithHeader("User-Agent", userAgent)
                                               .GetStringAsync(cancellationToken: cancellationToken);

                JObject json = JObject.Parse(response);
                if (json["error"] != null)
                    throw new InvalidOperationException($"OpenStreetMap geocode failed: {json["error"]}");

                string address = json["display_name"]?.ToString()
                    ?? throw new InvalidOperationException("display_name missing in response");
                return address;
            });
        }
    }
}

File: 1-starter_project_nestjs/libs/common/src/open-street-maps.service.ts
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import axios, { AxiosError } from 'axios';
import { IMapsService } from '@mytravels/contract';

const MAX_RETRIES = 2;
const DEFAULT_BASE_URL = 'https://nominatim.openstreetmap.org';
const DEFAULT_USER_AGENT = 'mytravels/1.0';

@Injectable()
export class OpenStreetMapsService implements IMapsService {
  private readonly baseUrl: string;
  private readonly userAgent: string;

  constructor(private readonly config: ConfigService) {
    this.baseUrl = config.get<string>('OPEN_STREET_MAPS_URL') ?? DEFAULT_BASE_URL;
    this.userAgent = config.get<string>('OPEN_STREET_MAPS_USER_AGENT') ?? DEFAULT_USER_AGENT;
  }

  async getAddress(latitude: number, longitude: number): Promise<string> {
    return this.withRetry(() =>
      axios.get(`${this.baseUrl}/reverse`, {
        params: { format: 'jsonv2', lat: latitude, lon: longitude },
        headers: { 'User-Agent': this.userAgent },
      }).then((res) => {
        if (res.data?.error) throw new Error(`OpenStreetMap geocode failed: ${res.data.error}`);
        const address = res.data?.display_name;
        if (!address) throw new Error('display_name missing in response');
        return address as string;
      }),
    );
  }

  private async withRetry<T>(fn: () => Promise<T>, attempt = 0): Promise<T> {
    try {
      return await fn();
    } catch (err) {
      if (attempt < MAX_RETRIES && err instanceof AxiosError) {
        const delay = Math.pow(2, attempt + 1) * 1000;
        await new Promise((r) => setTimeout(r, delay));
        return this.withRetry(fn, attempt + 1);
      }
      throw err;
    }
  }
}
